crates/kv-store/src/block.rs

```diff
@@ -0,0 +1,659 @@
+use std::{
+    fmt::Debug,
+    ops::{Bound, Range},
+    sync::Arc,
+};
+
+use bytes::{Buf, BufMut, Bytes, BytesMut};
+use loro_common::LoroResult;
+
+use crate::{compress::{compress, decompress, CompressionType}, iter::KvIterator, sstable::{get_common_prefix_len_and_strip,  SIZE_OF_U32, XXH_SEED}};
+
+use super::sstable::{ SIZE_OF_U16, SIZE_OF_U8};
+
+
+#[derive(Debug)]
+pub struct LargeValueBlock{
+    // without checksum
+    pub value_bytes: Bytes,
+    pub key: Bytes,
+}
+
+impl LargeValueBlock{
+    /// ┌──────────────────────────┐
+    /// │Large Block               │
+    /// │┌ ─ ─ ─ ┬ ─ ─ ─ ─ ─ ─ ─ ─ │
+    /// │  value   Block Checksum ││
+    /// ││ bytes │      u32        │
+    /// │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘│
+    /// └──────────────────────────┘
+    fn encode(&self, compression_type: CompressionType)->Bytes{
+        let mut buf = Vec::with_capacity(self.value_bytes.len() + SIZE_OF_U32);
+        buf.put_slice(&self.value_bytes);
+        if !compression_type.is_none(){
+            let compressed_data = compress(&buf, compression_type);
+            let checksum = xxhash_rust::xxh32::xxh32(&compressed_data, XXH_SEED);
+            buf.put_u32(checksum);
+        }else{
+            let checksum = xxhash_rust::xxh32::xxh32(&buf, XXH_SEED);
+            buf.put_u32(checksum);
+        }
+        buf.into()
+    }
+
+    fn decode(bytes:Bytes, key: Bytes, compression_type: CompressionType)->LoroResult<Self>{
+        let value_bytes = if compression_type.is_none(){
+            bytes.slice(..bytes.len() - SIZE_OF_U32)
+        }else{
+            decompress(bytes.slice(..bytes.len() - SIZE_OF_U32), compression_type)?
+        };
+        Ok(LargeValueBlock{
+            value_bytes,
+            key,
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct NormalBlock {
+    pub data: Bytes,
+    pub first_key: Bytes,
+    pub offsets: Vec<u16>,
+}
+
+impl NormalBlock {
+    /// ┌────────────────────────────────────────────────────────────────────────────────────────┐
+    /// │Block                                                                                   │
+    /// │┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ ┬ ─ ─ ─ ─┌ ─ ─ ─┌ ─ ─ ─ ┬ ─ ─ ─ ─┌ ─ ─ ─ ─ ─ ─ ─ │
+    /// │ Key Value Chunk  ...  │Key Value Chunk  offset │ ...  │ offset  kv len │Block Checksum││
+    /// ││     bytes     │      │     bytes     │  u16   │      │  u16  │  u16   │     u32       │
+    /// │ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ─ ┘─ ─ ─ ─ ─ ─ ─ ┘│
+    /// └────────────────────────────────────────────────────────────────────────────────────────┘
+    /// 
+    /// check sum will be calculated by crc32 later
+    fn encode(&self, compression_type: CompressionType) -> Bytes {
```

Comment:
It can be more efficient if it uses `&mut impl std::io::Write` to write to the target directly. 
We can avoid the allocations here